File: apps/api/app/models/schemas/analytics.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

ProductEventName = Literal[
    "paywall_viewed",
    "purchase_started",
    "purchase_succeeded",
    "purchase_failed",
    "push_permission",
]


class ProductEventIn(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: ProductEventName
    properties: dict[str, str] = Field(default_factory=dict)
    platform: Literal["ios", "android", "web"] | None = None
    app_version: str | None = Field(default=None, max_length=32)
    installation_id: str | None = Field(default=None, max_length=128)
    client_at: datetime | None = None

    @field_validator("properties")
    @classmethod
    def validate_property_size(cls, value: dict[str, str]) -> dict[str, str]:
        if len(value) > 8:
            raise ValueError("At most 8 event properties are allowed")
        if any(len(key) > 32 or len(item) > 64 for key, item in value.items()):
            raise ValueError("Event property keys or values are too long")
        return value


class ProductEventBatchIn(BaseModel):
    model_config = ConfigDict(extra="forbid")

    events: list[ProductEventIn] = Field(min_length=1, max_length=20)

File: apps/mobile/lib/api/analytics.ts
import { request } from "@/lib/api/client";

export type ProductEventName =
  | "paywall_viewed"
  | "purchase_started"
  | "purchase_succeeded"
  | "purchase_failed"
  | "push_permission";

export type ProductEventInput = {
  name: ProductEventName;
  properties: Record<string, string>;
  platform?: "ios" | "android" | "web";
  app_version?: string;
  installation_id?: string;
  client_at: string;
};

export const analyticsApi = {
  recordProductEvents: (token: string, events: ProductEventInput[]) =>
    request<void>("/analytics/events", token, {
      method: "POST",
      body: JSON.stringify({ events }),
    }),
};
